File: solution/go/detection/bad_domain.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strings"

	nlpb "github.com/bearlyrunning/FindingTheNeedle/solution/go/generated/normalizedlogpb"
	spb "github.com/bearlyrunning/FindingTheNeedle/solution/go/generated/signalpb"
	tspb "google.golang.org/protobuf/types/known/timestamppb"
)

const indicatorPath = "../../data/indicators/bad_domain.csv"

func (bdd *BadDomainDetection) ruleName() string {
	return bdd.name
}

func fmtRegex(ind []string) string {
	return fmt.Sprintf(".*(%s)$", strings.Join(ind, "|"))
}

func (bdd *BadDomainDetection) setFilterRegex() error {
	// Get the list of domain indicators.
	var ind []string
	f, err := os.Open(indicatorPath)
	if err != nil {
		return fmt.Errorf("failed to open file %s: %v", indicatorPath, err)
	}
	defer f.Close()
	s := bufio.NewScanner(f)
	for s.Scan() {
		ind = append(ind, strings.Split(s.Text(), ",")[0])
	}
	if err = s.Err(); err != nil {
		return fmt.Errorf("failed to scan file: %v", err)
	}
	// Compile a regex expression for matching indicators of compromise

	str := fmtRegex(ind)
	bdd.rr, err = regexp.Compile(str)
	if err != nil {
		return fmt.Errorf("failed compiling regex %s: %v", str, err)
	}
	return nil
}

// Filter out DNS logs containing known bad domains.
func (bdd *BadDomainDetection) filter() []*nlpb.DNS {
	var matched []*nlpb.DNS
	for _, log := range bdd.logs.dns {
		if bdd.rr.MatchString(log.GetQuery()) || bdd.rr.MatchString(log.GetAnswer()) {
			matched = append(matched, log)
		}
	}
	return matched
}

// Aggregate DNS logs containing known bad domains by source IP
// along with the time window of observed bad traffic.
func (bdd *BadDomainDetection) aggregate(matched []*nlpb.DNS) []*spb.Signal {
	var sigs []*spb.Signal
	keyBySrc := make(map[string][]*nlpb.DNS)
	for _, m := range matched {
		keyBySrc[m.SourceIp] = append(keyBySrc[m.SourceIp], m)
	}
	for src, logs := range keyBySrc {
		// Identify the time window of the bad DNS traffic.
		earliest, latest := logs[0].Timestamp.AsTime(), logs[0].Timestamp.AsTime()
		for _, l := range logs[1:] {
			if l.Timestamp.AsTime().Before(earliest) {
				earliest = l.Timestamp.AsTime()
				continue
			}
			if l.Timestamp.AsTime().After(latest) {
				latest = l.Timestamp.AsTime()
			}
		}

		// Find the bad indicator queried.
		m := bdd.rr.FindStringSubmatch(logs[0].GetQuery())
		if len(m) == 0 {
			m = bdd.rr.FindStringSubmatch(logs[0].GetAnswer())
		}

		sigs = append(sigs, &spb.Signal{
			Event: &spb.Signal_BadDomain{
				BadDomain: &spb.BadDomain{
					TimestampStart: tspb.New(earliest),
					TimestampEnd:   tspb.New(latest),
					SourceIp:       src,
					BadDomain:      m[1],
					DnsLog:         logs,
				},
			},
		})
	}
	return sigs
}

func (bdd *BadDomainDetection) run() ([]*spb.Signal, error) {
	// Set regex for filtering.
	if err := bdd.setFilterRegex(); err != nil {
		return nil, err
	}
	// Run detection logic: 1) filter 2) aggregate.
	return bdd.aggregate(bdd.filter()), nil
}

File: solution/python/detection/bad_domain.py
import detection
import logging
import re
from generated import normalized_log_pb2 as nlpb
from generated import signal_pb2 as spb
import google.protobuf.timestamp_pb2 as tspb

INDICATOR_PATH = "../../data/indicators/bad_domain.csv"

class BadDomainDetection(detection.DetectionInterface):
    def __init__(self, name="", logs=[], indicators=[], signals=[]):
        self.name = name
        self.logs = logs
        self.indicators = indicators
        self.signals = signals
        self.regStr = ""

    # Get detection rule name.
    def ruleName(self) -> str:
        return self.name

    # Provide a regular expression string based on detection indicators.
    def fmtRegex(self):
        self.regStr = ".*(" + "|".join(self.indicators) + ")$"
    
    # Sets rule indicators based on INDICATOR_PATH.
    def setFilterRegex(self) -> None:
        # Open indicators
        try:
            lst = open(INDICATOR_PATH)
        except Exception as e:
             logging.error("failed to open file %s: %s", INDICATOR_PATH, str(e))
             raise e
        
        for indicator in lst:
            self.indicators.append(indicator.strip().split(",")[0])

        lst.close()

    # Match indicator presence in logs.
    def filter(self) -> list[nlpb.DNS]:
        self.setFilterRegex
        self.fmtRegex()
        matched = []
        
        for log in self.logs.dns:
            if re.search(self.regStr, log.query) or re.search(self.regStr, log.answer):
                matched.append(log)

        return matched
    
    # Comput signal aggregation based on source IP address.
    def aggregate(self, matched=[]) -> list[spb.Signal]:
        sigs = []
        # Initialise a dict of source_ip -> list[nlpb.DNS]
        keyBySrc = {}

        # Aggregate by source_ip.
        for m in matched:
            if m.source_ip not in keyBySrc:
                keyBySrc[m.source_ip] = [m]
            else:
                keyBySrc[m.source_ip].append(m)

        for src, logs in keyBySrc.items():
            # Determine time window of aggregated logs.
            earliest = logs[0].timestamp.ToDatetime()
            latest = logs[0].timestamp.ToDatetime()
            for l in logs[1:]:
                if l.timestamp.ToDatetime() < earliest:
                    earliest = l.timestamp.ToDatetime()
                    continue
                if l.timestsamp.ToDateTime() > latest:
                    latest = l.timestamp.ToDatetime()
            ts_earliest = tspb.Timestamp()
            ts_earliest.FromDatetime(earliest)
            ts_latest = tspb.Timestamp()
            ts_latest.FromDatetime(latest)

            # Find the bad indicator quried
            m = re.findall(self.regStr, logs[0].query)
            if len(m) == 0:
                m = re.findall(self.regStr, logs[0].answer)

            # Create aggregated entry per source_ip.
            sigs.append(spb.Signal(
                bad_domain=spb.BadDomain(
                    timestamp_start=ts_earliest,
                    timestamp_end=ts_latest,
                    source_ip=src,
                    bad_domain=m[0],
                    dns_log=logs
                )
            ))

        return sigs

    # Run detection logic.
    def run(self) -> list[spb.Signal]:
        # Set regex for filtering.
        try:
            self.setFilterRegex()
        except Exception as e:
            raise e
        
        # Find any logs that contain indicators of compromise.
        # Run detection logic: 1) filter 2) aggregate.
        return self.aggregate(self.filter())